### PR TITLE
Kernels community fa3

### PR DESCRIPTION
**Pull request by rainj-me**
## Motivation

Support flash-attn-3 from kernels community

## Modifications

- Use the fa3 kernel from kernels community
- Docker image
- CI

## Accuracy Tests

```
100%|████████████████████████████████████████████| 100/100 [00:05<00:00, 17.00it/s]
Accuracy: 0.950
Invalid: 0.000
Latency: 5.882 s
Output throughput: 2046.968 token/s
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/jit_kernel/flash_attention_v3.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+from typing import Optional, Union
+
+import torch
+
+from sglang.jit_kernel.utils import cache_once
+
+logger = logging.getLogger(__name__)
+
+SGL_FA3_KERNEL_REPO = "kernels-community/sgl-flash-attn3"
+DEFAULT_FA3_KERNEL_LOCKFILE = "/sgl-workspace/sglang/python/kernels.lock"
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/flash_attention_v3.py]: Why hardcode an absolute path here? That's not portable

Comment (DarkSharpness) [reply python/sglang/jit_kernel/flash_attention_v3.py]: You may try `~/.cache/sglang/fa3.lock` or something similar.

File: python/sglang/srt/layers/attention/xpu_backend.py [hunk 2]
```diff
@@ -497,7 +503,7 @@ def forward_extend(
                 cu_seqlens_k = metadata.encoder_cu_seqlens_k
                 window_size = (-1, -1)
 
-            result = flash_attn_with_kvcache(
+            result = flash_attn_with_kvcache_fa3(
```
Comment (DarkSharpness) [inline python/sglang/srt/layers/attention/xpu_backend.py]: I'm not sure whether we keep the original naming `flash_attn_with_kvcache` here? Given that only fa3 is supported on this platform, maybe simply using `flash_attn_with_kvcache` is ok.

File: python/sglang/jit_kernel/flash_attention_v3.py (hunk 1 above)
Comment (DarkSharpness) [reply python/sglang/jit_kernel/flash_attention_v3.py]: If the lock must be within the SGLang package, please use absolute path relative to this file to find the lock.

File: scripts/ci/cuda/ci_install_dependency.sh [hunk 3]
```diff
@@ -333,6 +333,9 @@ bash "${SCRIPT_DIR}/ci_download_flashinfer_cubin.sh"
 $PIP_UNINSTALL_CMD nvidia-cutlass-dsl-libs-base $PIP_UNINSTALL_SUFFIX || true
 $PIP_CMD install nvidia-cutlass-dsl==4.3.5 --force-reinstall $PIP_INSTALL_SUFFIX || true
 
+# Download kernels from kernels community
+kernels download python || true
```
Comment (DarkSharpness) [inline scripts/ci/cuda/ci_install_dependency.sh:362]: What does this mean? Will this install all the kernels? We should only add required kernels in our dependency.

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 4]
```diff
@@ -61,7 +61,12 @@
             "aiter is AMD specific kernel library. Please make sure aiter is installed on your AMD device."
         )
 else:
-    from sgl_kernel.flash_attn import flash_attn_varlen_func, flash_attn_with_kvcache
+    from sglang.jit_kernel.flash_attention_v3 import (
+        flash_attn_varlen_func,
+    )
+    from sglang.jit_kernel.flash_attention_v3 import (
```
Comment (DarkSharpness) [inline python/sglang/srt/layers/attention/nsa_backend.py]: Similarly, I'm not sure whether we keep the original naming flash_attn_with_kvcache here? Given that only fa3 is supported on this platform, maybe simply using flash_attn_with_kvcache is ok.

File: python/sglang/srt/layers/attention/xpu_backend.py (hunk 2 above)
Comment (rainj-me) [reply python/sglang/srt/layers/attention/xpu_backend.py]: Sure will revert the change.

File: python/sglang/jit_kernel/flash_attention_v3.py (hunk 1 above)
Comment (rainj-me) [reply python/sglang/jit_kernel/flash_attention_v3.py]: I will try to remove the lock file from the repo. And only load it when provide from a default path like **~/.cache/sglang/kernel.lock**

File: scripts/ci/cuda/ci_install_dependency.sh (hunk 3 above)
Comment (rainj-me) [reply scripts/ci/cuda/ci_install_dependency.sh:362]: This command will only download the kernerls specificed in pyproject.toml

```
[tool.kernels.dependencies]
"kernels-community/sgl-flash-attn3" = 1
```

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 4 above)
Comment (rainj-me) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Since we are using the fa4 official pkg, the fa4 support on sm90 is possible. rename this is trying to make the refactoring in future easier.

File: python/sglang/jit_kernel/flash_attention_v3.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+@cache_once
+def _load_fa3_kernels():
+    lockfile_path = os.path.join(
+        envs.SGLANG_CACHE_DIR.get(), DEFAULT_FA3_KERNEL_LOCKFILE
+    )
+    try:
+        # When the lock file provided, load from the kernel cache directory,
+        # otherwise, load from the repo, which require download from huggingface hub
+        # but always works as long as the repo is accessible.
+        if os.path.exists(DEFAULT_FA3_KERNEL_LOCKFILE):
```
Comment (adarshxs) [inline python/sglang/jit_kernel/flash_attention_v3.py]: we should check `lockfile_path` instead of `DEFAULT_FA3_KERNEL_LOCKFILE`?

Comment (rainj-me) [reply python/sglang/jit_kernel/flash_attention_v3.py]: good catch, will fix asap

File: .gitignore [hunk 6]
```diff
@@ -257,6 +257,7 @@ inputs/
 
 # setuptools-scm generated version file
 python/sglang/_version.py
+python/kernel.lock
```
Comment (DarkSharpness) [inline .gitignore:261]: Do we still need this? Or maybe just remove it.

File: python/sglang/jit_kernel/flash_attention_v3.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        scheduler_metadata,
+        num_splits,
+        pack_gqa,
+        sm_margin,
+        return_softmax_lse,
+        sinks,
+    )
+
+
+def flash_attn_varlen_func(
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/flash_attention_v3.py:165]: Can we provide a unified interface for `flash_attn_varlen_func` so that we can seaminglessly switch from v3 to v4 through a kwargs `version`?

File: .gitignore (hunk 6 above)
Comment (rainj-me) [reply .gitignore:261]: This is .gitignore file, I'm trying to make sure it won't commit to repo.

File: python/sglang/jit_kernel/flash_attention_v3.py (hunk 7 above)
Comment (rainj-me) [reply python/sglang/jit_kernel/flash_attention_v3.py:165]: Sure, let me refactor it.

File: python/sglang/jit_kernel/flash_attention.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+    )
+    from .flash_attention_v4 import flash_attn_varlen_func as fa4_flash_attn_varlen_func
+    from .flash_attention_v4 import (
+        flash_attn_with_kvcache as fa4_flash_attn_with_kvcache,
+    )
+except ImportError as e:
+    raise e
+
+
+class FlashAttnVersion(Enum):
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/flash_attention.py]: Do we really need enum? I think simply using int is ok here.

File: python/sglang/jit_kernel/flash_attention.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
+from enum import Enum
+from typing import Optional, Union
+
+import torch
+
+# TODO: the following import may cause error
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/flash_attention.py]: What may cause error? BTW, this try except is redundant. Remove it.

Comment (rainj-me) [reply python/sglang/jit_kernel/flash_attention.py]: Ah, good catch, I just forgot to remove the comments.

File: python/sglang/jit_kernel/flash_attention.py (hunk 8 above)
Comment (rainj-me) [reply python/sglang/jit_kernel/flash_attention.py]: Sure.

File: python/sglang/jit_kernel/flash_attention_v3.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+SGL_FA3_KERNEL_REVISION = "v1"
+DEFAULT_FA3_KERNEL_LOCKFILE = "kernels.lock"
+
+
+@cache_once
+def _load_fa3_kernels():
+    lockfile_path = os.path.join(
+        envs.SGLANG_CACHE_DIR.get(), DEFAULT_FA3_KERNEL_LOCKFILE
+    )
+    try:
```
Comment (Fridge003) [inline python/sglang/jit_kernel/flash_attention_v3.py:33]: Can user manually choose between huggingface fa3 and sgl-kernel fa3?
Since they might have diffs

File: python/sglang/jit_kernel/flash_attention_v3.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
+import logging
```
Comment (Fridge003) [inline python/sglang/jit_kernel/flash_attention_v3.py:1]: Why are we changing the entrance to jit kernel?
Since fa3/fa4 are not jit kernels

File: docker/Dockerfile [hunk 12]
```diff
@@ -219,6 +219,9 @@ RUN --mount=type=cache,target=/root/.cache/pip \
         python3 -m pip install flashinfer-jit-cache==${FLASHINFER_VERSION} --index-url https://flashinfer.ai/whl/cu${CUINDEX} ; \
     fi \
     && FLASHINFER_CUBIN_DOWNLOAD_THREADS=${BUILD_AND_DOWNLOAD_PARALLEL} FLASHINFER_LOGGING_LEVEL=warning python3 -m flashinfer --download-cubin
+    && kernels download python
+    && kernels lock python
+    && mv python/kernels.lock /root/.cache/sglang
```
Comment (Fridge003) [inline docker/Dockerfile:224]: Avoid hardcoding of lock path

File: docs/references/environment_variables.md [hunk 13]
```diff
@@ -19,6 +19,7 @@ SGLang supports various environment variables that can be used to configure its
 | `SGLANG_FORWARD_UNKNOWN_TOOLS`            | Forward unknown tool calls to clients instead of dropping them                                                                   | `false` (drop unknown tools) |
 | `SGLANG_REQ_WAITING_TIMEOUT`              | Timeout (in seconds) for requests waiting in the queue before being scheduled                                                    | `-1`                         |
 | `SGLANG_REQ_RUNNING_TIMEOUT`              | Timeout (in seconds) for requests running in the decode batch                                                                    | `-1`                         |
+| `SGLANG_CACHE_DIR`                        | Cache directory for model weights and other data | `~/.cache/sglang` |
```
Comment (Fridge003) [inline docs/references/environment_variables.md:22]: We had better rename this environ, since it's only for storing lock from kernels

Comment (rainj-me) [reply docs/references/environment_variables.md:22]: this is a environment variable not created by me, it is created by piecewise cudagraph but forget to document.

File: docker/Dockerfile (hunk 12 above)
Comment (rainj-me) [reply docker/Dockerfile:224]: For container image, I believe we have to hard code, otherwise it will not work.

**Review by Fridge003 — APPROVED** [review]

**Comment by adarshxs** [comment]
@rainj-me we can add `kernels` to `pyproject_cpu.toml` maybe? or put the import behind a try/except

**Comment by jianan-gu** [comment]
@rainj-me    Hi, this PR breaks Xeon CI by the changes on test/srt/cpu/test_flash_attn.py, could you help fix? just revert the changes in test/srt/cpu/test_flash_attn.py    cc @Fridge003

**Comment by blzheng** [comment]
> @rainj-me Hi, this PR breaks Xeon CI by the changes on test/srt/cpu/test_flash_attn.py, could you help fix? just revert the changes in test/srt/cpu/test_flash_attn.py cc @Fridge003

I have resolved the Xeon CI failure in this PR: https://github.com/sgl-project/sglang/pull/21367

**Comment by rainj-me** [comment]
> > @rainj-me Hi, this PR breaks Xeon CI by the changes on test/srt/cpu/test_flash_attn.py, could you help fix? just revert the changes in test/srt/cpu/test_flash_attn.py cc @Fridge003
> 
> 
> 
> I have resolved the Xeon CI failure in this PR: https://github.com/sgl-project/sglang/pull/21367 

approved
